### PR TITLE
Feat:performance graphql context integration

### PR DESCRIPTION
**Pull request by harsh04044**
**What kind of change does this PR introduce?**

Feature - Performance monitoring integration with GraphQL context

**Issue Number:**

Fixes #4125 

**Snapshots/Videos:**

N/A - Backend API changes only

**If relevant, did you update the documentation?**

Yes - Updated performance monitoring documentation to reflect GraphQL context integration.
**Summary**

This PR is **PR 2 of 8-9 planned PRs** in the performance monitoring and observability enhancement series. It integrates the existing performance monitoring system with GraphQL operations, enabling comprehensive performance tracking for all GraphQL queries, mutations, and subscriptions. Previously, the performance tracker was only available at the HTTP request level. This change makes it accessible within GraphQL resolvers through the context object, allowing developers to track operation-specific metrics and GraphQL query complexity scores.

**Key Changes:**
- Added optional `perf` field to `ExplicitGraphQLContext` type, making the performance tracker available in all GraphQL resolvers
- Integrated performance tracker in `createContext` function for HTTP-based GraphQL requests
- Implemented GraphQL complexity score tracking in the `preExecution` hook, capturing query complexity after mutation base costs are applied
- Added support for extracting performance tracker from WebSocket connections in subscription `onConnect` handler
- Created comprehensive test coverage for performance plugin and GraphQL integration
- Fixed documentation typo: corrected `cacheMiss` to `cacheMisses` to match the actual `PerfSnapshot` type

This integration maintains backward compatibility (the `perf` field is optional) and follows the existing patterns established by the performance monitoring foundation.

**Does this PR introduce a breaking change?**

No - This change is fully backward compatible. The `perf` field is optional in the GraphQL context, so existing code will continue to work without modification. Resolvers that don't use the performance tracker will function exactly as before.

### CodeRabbit AI Review
- [ ] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [ ] I have implemented or provided justification for each non-critical suggestion
- [ ] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [x] I have written tests for all new changes/features
- [x] I have verified that test coverage meets or exceeds 95%
- [x] I have run the test suite locally and all tests pass

**Test Details:**
- Created comprehensive test suite for performance plugin (`test/fastifyPlugins/performance.test.ts`) covering:
  - Plugin registration and endpoint availability
  - Request hook behavior (perf tracker attachment, timestamp tracking)
  - Response hook behavior (Server-Timing headers, snapshot storage)
  - Edge cases (missing perf tracker, concurrent requests, no cache/DB operations)
- Extended GraphQL route tests (`test/routes/graphql.test.ts`) with integration tests for:
  - Context creation with/without perf tracker
  - Complexity score tracking in preExecution hook
  - Subscription onConnect perf tracker extraction
  - All edge cases (undefined socket, missing request, invalid perf objects)
- All 12 test shards pass successfully
- 100% code coverage for all new code paths

**Other information**

**Technical Implementation Notes:**
- The performance tracker is attached to requests via the performance plugin's `onRequest` hook
- For HTTP requests, the tracker is directly accessible via `request.perf` and passed to GraphQL context
- For WebSocket subscriptions, the tracker is extracted from `socket.request.perf` using a type guard to ensure type safety
- Complexity tracking happens in the `preExecution` hook after mutation base costs are calculated, ensuring accurate complexity scores
- Socket extraction uses optional chaining (`socket?.request`) for optimal performance and readability

**Files Changed:**
- `src/graphql/context.ts` - Added `perf` field to `ExplicitGraphQLContext`
- `src/routes/graphql.ts` - Integrated perf tracker in context creation, complexity tracking, and subscription handling
- `test/fastifyPlugins/performance.test.ts` - New comprehensive test file
- `test/routes/graphql.test.ts` - Extended with GraphQL performance integration tests
- `docs/docs/docs/developer-resources/best-practices/performance-monitoring.md` - Fixed documentation typo
- Auto-generated documentation files updated

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

## Summary by CodeRabbit

* **New Features**
  * Per-request performance tracking for GraphQL (HTTP + subscriptions) with complexity reporting
  * HTTP responses emit Server-Timing headers (DB, cache, total)
  * Added runtime detection/validation for performance tracker objects
  * /metrics/perf endpoint returning recent snapshots (newest first, capped)

* **Tests**
  * Extensive tests for performance plugin, GraphQL perf integration, and perf-type detection

* **Chores**
  * Removed duplicate CI configuration entry

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by harsh04044** [comment]
@palisadoes  review this

**Comment by palisadoes** [comment]
@coderabbitai full review based on your configuration. Post the results as a comment.

**Review by palisadoes — APPROVED** [review]

**Comment by palisadoes** [comment]
Please:

1. Fix any failing tests
1. Ensure test code coverage for your patch meets our minimum >=95% requirements
1. Fix any CI/CD failures. Click on the `Details` link for the tests to get an explanation of the cause(s)
    1. The sensitive files check does not prevent the entire test suite from running. 
1. Ensure code rabbit approves your changes
1. Remedy any conflicting files

File: src/utilities/metrics/performanceTracker.ts [hunk 1]
```diff
@@ -261,3 +261,22 @@ export function createPerformanceTracker(
...
+	return (
+		typeof value === "object" &&
+		value !== null &&
+		"trackComplexity" in value &&
+		"snapshot" in value &&
+		"trackDb" in value &&
+		"trackCacheHit" in value &&
+		"trackCacheMiss" in value
+	);
+}
```
[the remark this answers is not in the document]
Comment (harsh04044) [reply src/utilities/metrics/performanceTracker.ts:284]: i have addresed this in my last commit check

**Comment by harsh04044** [comment]
@coderabbitai full review . Post the results as a comment.

**Comment by harsh04044** [comment]
@coderabbitai Full review . Post the results as a comment.

File: test/routes/graphql.test.ts [hunk 2]
```diff
@@ -429,6 +432,75 @@ describe("GraphQL Routes", () => {
...
+				expect(context.perf).toBeDefined();
+				expect(context.perf).toBe(perfTracker);
+				expect(context.perf).toHaveProperty("trackComplexity");
+				expect(context.perf).toHaveProperty("snapshot");
+
+				// Verify cache is wrapped with metrics proxy
+				expect(context.cache).not.toBe(mockFastify.cache);
+				expect(context.cache).toHaveProperty("get");
+				expect(context.cache).toHaveProperty("set");
+			});
```
[the remark this answers is not in the document]
Comment (harsh04044) [reply test/routes/graphql.test.ts:466]: i have resolved it in my latest commit

File: test/routes/graphql.test.ts [hunk 3]
```diff
@@ -1097,6 +1249,166 @@ describe("GraphQL Routes", () => {
...
+				socket,
+			});
+
+			expect(result).toEqual(
+				expect.objectContaining({
+					currentClient: { isAuthenticated: true, user: decoded.user },
+					perf: undefined,
+				}),
+			);
+		});
```
[the remark this answers is not in the document]
Comment (harsh04044) [reply test/routes/graphql.test.ts:1416]: this is resolved

**Comment by harsh04044** [comment]
@coderabbitai Full review . Post the results as a comment.

**Comment by harsh04044** [comment]
@coderabbitai What is the conclusion from your analysis here?:

**Comment by harsh04044** [comment]
@palisadoes  review this

**Comment by palisadoes** [comment]
@coderabbitai full review

**Review by palisadoes — APPROVED** [review]

**Comment by palisadoes** [comment]
@coderabbitai What is the conclusion of your assessment here?:
- https://github.com/PalisadoesFoundation/talawa-api/pull/4367#issuecomment-3736615387
